stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1385 @@
+use super::stacks::{StacksBlockHeader, StacksMicroblock, TransactionPayload};
+use crate::burnchains::PoxConstants;
+use crate::chainstate::burn::db::sortdb::SortitionDB;
+use crate::chainstate::stacks::db::StacksChainState;
+use crate::chainstate::stacks::{MINER_BLOCK_CONSENSUS_HASH, MINER_BLOCK_HEADER_HASH};
+use crate::clarity_vm::clarity::{ClarityInstance, PreCommitClarityBlock};
+use crate::clarity_vm::database::SortitionDBRef;
+use crate::monitoring;
+use crate::util_lib::db::{query_row_panic, query_rows, u64_to_sql, Error as DBError, FromRow};
+
+#[cfg(test)]
+pub mod tests;
+
+define_named_enum!(HeaderTypeNames {
+    Nakamoto("nakamoto"),
+    Epoch2("epoch2"),
+});
+
+impl ToSql for HeaderTypeNames {
+    fn to_sql(&self) -> rusqlite::Result<rusqlite::types::ToSqlOutput<'_>> {
+        self.get_name_str().to_sql()
+    }
+}
+
+impl FromSql for HeaderTypeNames {
+    fn column_result(value: rusqlite::types::ValueRef<'_>) -> rusqlite::types::FromSqlResult<Self> {
+        Self::lookup_by_name(value.as_str()?).ok_or_else(|| FromSqlError::InvalidType)
+    }
+}
+
+lazy_static! {
+    pub static ref NAKAMOTO_CHAINSTATE_SCHEMA_1: Vec<String> = vec![
+    r#"
+      -- Table for Nakamoto Block Headers
+      CREATE TABLE nakamoto_block_headers (
+          -- The following fields all correspond to entries in the StacksHeaderInfo struct
+                     block_height INTEGER NOT NULL,
+                     -- root hash of the internal, not-consensus-critical MARF that allows us to track chainstate/fork metadata
+                     index_root TEXT NOT NULL,
+                     -- all consensus hashes are guaranteed to be unique
+                     consensus_hash TEXT UNIQUE NOT NULL,
+                     -- burn header hash corresponding to the consensus hash (NOT guaranteed to be unique, since we can 
+                     --    have 2+ blocks per burn block if there's a PoX fork)
+                     burn_header_hash TEXT NOT NULL,
+                     -- height of the burnchain block header that generated this consensus hash
+                     burn_header_height INT NOT NULL,
+                     -- timestamp from burnchain block header that generated this consensus hash
+                     burn_header_timestamp INT NOT NULL,
+                     block_size TEXT NOT NULL,
+          -- The following fields all correspond to entries in the NakamotoBlockHeader struct
+                     version INTEGER NOT NULL,
+                     -- this field is the total number of blocks in the chain history (including this block)
+                     chain_length INTEGER NOT NULL,
+                     -- this field is the total amount of BTC spent in the chain history (including this block)
+                     burn_spent INTEGER NOT NULL,
+                     -- the parent BlockHeaderHash
+                     parent TEXT NOT NULL,
+                     -- the latest bitcoin block whose data is viewable from this stacks block
```

Comment:
Is this a `BurnHeaderHash`?  How is this different from `burn_header_hash` above?

Reply:
One is the bitcoin block for which data has been processed and can be queried by clarity. The other is the bitcoin block in which the tenure was elected. They do not necessarily have to be the same (consider the case of a missed sortition).

Reply:
I'm still not seeing this.  A Nakamoto block header contains the hash of the Bitcoin block in which its tenure began.  If a sortition is missed, then Stackers grant the currently-active miner a new tenure (a new `TenureChange` transaction), which ought to cause the `burn_view` field to change to the next Bitcoin block (i.e. the one without the sortition).

Reply:
> I'd argue that Nakamoto Block C's "consensus_hash" should still correspond to Bitcoin Block 10.

That's where we differ.  I'd argue that Nakamoto block C's consensus hash should be the same as Bitcoin block 11's consensus hash.  This is because it vastly simplifies evaluating `(get-burn-block-info?)`:  to query the burnchain state in the `BurnStateDB` implementation, we simply query the sortition history tipped by the consensus hash in the transaction's block.  This is only possible if Nakamoto block C's consensus hash is derived from tenure in which it is mined (i.e. Bitcoin block 11), instead of the miner that mined it.

We don't do this yet, but I think we'll want to have a `tenures` table in the `NakamotoChainState` that keeps track of all tenures that have been executed (even if there was no corresponding sortition, and even if there were no Stacks blocks mined in the tenure).  That is, we need to treat tenures as first-class objects in Stacks.  If we tracked them independent of block headers, then we could do things like query the current tenure's miner (and thus its public key) as well as the tenure's corresponding sortition (so we can get the PoX payouts).  In addition, it would make it easy to select all Stacks blocks by tenure, so we can serve them via the RPC endpoint on boot-up.  We have to do all of this anyway because the design calls for one `TenureChange` transaction per Bitcoin block, and having a separate table describing them all would greatly simplify this.

Reply:
I can add a `tenures` table in my coordinator PR